In machine learning you will most likely use databases from many sources to train your learning algorithms. Pandas allows us to load databases of different formats into DataFrames. One of the most popular data formats used to store databases is csv. CSV stands for Comma Separated Values and offers a simple format to store data. We can load CSV files into Pandas DataFrames using the ```pd.read_csv()``` function. Let's load Google stock data into a Pandas DataFrame. The GOOG.csv file contains Google stock data from 8/19/2004 till 10/13/2017 taken from Yahoo Finance.

In [2]:
import pandas as pd

In [28]:
data = pd.read_csv("/content/drive/MyDrive/devtown-10922/DevTown/Python-for-Data-Science--master/stocks.csv")
data

Date        Open        High         Low       Close   Adj Close  \
0     2004-08-19   49.676899   51.693783   47.669952   49.845802   49.845802   
1     2004-08-20   50.178635   54.187561   49.925285   53.805050   53.805050   
2     2004-08-23   55.017166   56.373344   54.172661   54.346527   54.346527   
3     2004-08-24   55.260582   55.439419   51.450363   52.096165   52.096165   
4     2004-08-25   52.140873   53.651051   51.604362   52.657513   52.657513   
...          ...         ...         ...         ...         ...         ...   
3308  2017-10-09  980.000000  985.424988  976.109985  977.000000  977.000000   
3309  2017-10-10  980.000000  981.570007  966.080017  972.599976  972.599976   
3310  2017-10-11  973.719971  990.710022  972.250000  989.250000  989.250000   
3311  2017-10-12  987.450012  994.119995  985.000000  987.830017  987.830017   
3312  2017-10-13  992.000000  997.210022  989.000000  989.679993  989.679993   

        Volume  
0     44994500  
1     23005800  
2     18393200  
3     15361800  
4      9257400  
...        ...  
3308    891400  
3309    968400  
3310   1693300  
3311   1262400  
3312   1157700  

[3313 rows x 7 columns]

In [29]:
print(type(data))
print(data.shape)

<class 'pandas.core.frame.DataFrame'>
(3313, 7)


In [30]:
import pandas as pd

# We load Google stock data in a DataFrame
Google_stock = pd.read_csv('/content/drive/MyDrive/DevTown/Python-for-Data-Science--master/stocks.csv')

# We print some information about Google_stock
print('Google_stock is of type:', type(Google_stock))
print('Google_stock has shape:', Google_stock.shape)

FileNotFoundError: ignored

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

We see that we have loaded the stocks.csv file into a Pandas DataFrame and it consists of 3,313 rows and 7 columns. Now let's look at the stock data

In [31]:
Google_stock

Date        Open        High         Low       Close   Adj Close  \
0     2004-08-19   49.676899   51.693783   47.669952   49.845802   49.845802   
1     2004-08-20   50.178635   54.187561   49.925285   53.805050   53.805050   
2     2004-08-23   55.017166   56.373344   54.172661   54.346527   54.346527   
3     2004-08-24   55.260582   55.439419   51.450363   52.096165   52.096165   
4     2004-08-25   52.140873   53.651051   51.604362   52.657513   52.657513   
...          ...         ...         ...         ...         ...         ...   
3308  2017-10-09  980.000000  985.424988  976.109985  977.000000  977.000000   
3309  2017-10-10  980.000000  981.570007  966.080017  972.599976  972.599976   
3310  2017-10-11  973.719971  990.710022  972.250000  989.250000  989.250000   
3311  2017-10-12  987.450012  994.119995  985.000000  987.830017  987.830017   
3312  2017-10-13  992.000000  997.210022  989.000000  989.679993  989.679993   

        Volume  
0     44994500  
1     23005800  
2     18393200  
3     15361800  
4      9257400  
...        ...  
3308    891400  
3309    968400  
3310   1693300  
3311   1262400  
3312   1157700  

[3313 rows x 7 columns]

We see that it is quite a large dataset and that Pandas has automatically assigned numerical row indices to the DataFrame. Pandas also used the labels that appear in the data in the CSV file to assign the column labels.

When dealing with large datasets like this one, it is often useful just to take a look at the first few rows of data instead of the whole dataset. We can take a look at the first 5 rows of data using the ```.head()``` method, as shown below

In [32]:
data.head()

Date       Open       High        Low      Close  Adj Close    Volume
0  2004-08-19  49.676899  51.693783  47.669952  49.845802  49.845802  44994500
1  2004-08-20  50.178635  54.187561  49.925285  53.805050  53.805050  23005800
2  2004-08-23  55.017166  56.373344  54.172661  54.346527  54.346527  18393200
3  2004-08-24  55.260582  55.439419  51.450363  52.096165  52.096165  15361800
4  2004-08-25  52.140873  53.651051  51.604362  52.657513  52.657513   9257400

We can also take a look at the last 5 rows of data by using the ```.tail()``` method:

In [33]:
data.tail()

Date        Open        High         Low       Close   Adj Close  \
3308  2017-10-09  980.000000  985.424988  976.109985  977.000000  977.000000   
3309  2017-10-10  980.000000  981.570007  966.080017  972.599976  972.599976   
3310  2017-10-11  973.719971  990.710022  972.250000  989.250000  989.250000   
3311  2017-10-12  987.450012  994.119995  985.000000  987.830017  987.830017   
3312  2017-10-13  992.000000  997.210022  989.000000  989.679993  989.679993   

       Volume  
3308   891400  
3309   968400  
3310  1693300  
3311  1262400  
3312  1157700

We can also optionally use ```.head(N)``` or ```.tail(N)``` to display the first and last N rows of data, respectively.

Let's do a quick check to see whether we have any ```NaN``` values in our dataset. To do this, we will use the ```.isnull()``` method followed by the ```.any()``` method to check whether any of the columns contain ```NaN``` values.

In [34]:
data.isnull().sum()

Date         0
Open         0
High         0
Low          0
Close        0
Adj Close    0
Volume       0
dtype: int64

We see that we have no ```NaN``` values.

When dealing with large datasets, it is often useful to get statistical information from them. Pandas provides the ```.describe()``` method to get descriptive statistics on each column of the DataFrame. Let's see how this works:

In [35]:
# We get descriptive statistics on our stock data
data.describe()

Open         High          Low        Close    Adj Close  \
count  3313.000000  3313.000000  3313.000000  3313.000000  3313.000000   
mean    380.186092   383.493740   376.519309   380.072458   380.072458   
std     223.818650   224.974534   222.473232   223.853780   223.853780   
min      49.274517    50.541279    47.669952    49.681866    49.681866   
25%     226.556473   228.394516   224.003082   226.407440   226.407440   
50%     293.312286   295.433502   289.929291   293.029114   293.029114   
75%     536.650024   540.000000   532.409973   536.690002   536.690002   
max     992.000000   997.210022   989.000000   989.679993   989.679993   

             Volume  
count  3.313000e+03  
mean   8.038476e+06  
std    8.399521e+06  
min    7.900000e+03  
25%    2.584900e+06  
50%    5.281300e+06  
75%    1.065370e+07  
max    8.276810e+07

If desired, we can apply the ```.describe()``` method on a single column as shown below:

In [36]:
# We get descriptive statistics on a single column of our DataFrame
data['Adj Close'].describe()

count    3313.000000
mean      380.072458
std       223.853780
min        49.681866
25%       226.407440
50%       293.029114
75%       536.690002
max       989.679993
Name: Adj Close, dtype: float64

Similarly, you can also look at one statistic by using one of the many statistical functions Pandas provides. Let's look at some examples:

In [37]:
# We print information about our DataFrame  
print()
print('Maximum values of each column:\n', data.max())
print()
print('Minimum Close value:', data['Close'].min())
print()
print('Average value of each column:\n', data.mean())


Maximum values of each column:
 Date         2017-10-13
Open              992.0
High         997.210022
Low               989.0
Close        989.679993
Adj Close    989.679993
Volume         82768100
dtype: object

Minimum Close value: 49.681866

Average value of each column:
 Open         3.801861e+02
High         3.834937e+02
Low          3.765193e+02
Close        3.800725e+02
Adj Close    3.800725e+02
Volume       8.038476e+06
dtype: float64


<ipython-input-37-151d4cdfd5fe>:7: FutureWarning: Dropping of nuisance columns in DataFrame reductions (with 'numeric_only=None') is deprecated; in a future version this will raise TypeError.  Select only valid columns before calling the reduction.
  print('Average value of each column:\n', data.mean())


In [38]:
data.std()

<ipython-input-38-a47ac8255c06>:1: FutureWarning: Dropping of nuisance columns in DataFrame reductions (with 'numeric_only=None') is deprecated; in a future version this will raise TypeError.  Select only valid columns before calling the reduction.
  data.std()


Open         2.238187e+02
High         2.249745e+02
Low          2.224732e+02
Close        2.238538e+02
Adj Close    2.238538e+02
Volume       8.399521e+06
dtype: float64

Another important statistical measure is data correlation. Data correlation can tell us, for example, if the data in different columns are correlated. We can use the .corr() method to get the correlation between different columns, as shown below:

In [26]:
# We display the correlation between columns
data.corr()

Open      High       Low     Close  Adj Close    Volume
Open       1.000000  0.999904  0.999845  0.999745   0.999745 -0.564258
High       0.999904  1.000000  0.999834  0.999868   0.999868 -0.562749
Low        0.999845  0.999834  1.000000  0.999899   0.999899 -0.567007
Close      0.999745  0.999868  0.999899  1.000000   1.000000 -0.564967
Adj Close  0.999745  0.999868  0.999899  1.000000   1.000000 -0.564967
Volume    -0.564258 -0.562749 -0.567007 -0.564967  -0.564967  1.000000

A correlation value of 1 tells us there is a high correlation and a correlation of 0 tells us that the data is not correlated at all.

We will end this Introduction to Pandas by taking a look at the .groupby() method. The .groupby() method allows us to group data in different ways. Let's see how we can group data to get different types of information. For the next examples we are going to load fake data about a fictitious company.

In [40]:
housing = pd.read_csv("/content/sample_data/california_housing_train.csv")
housing

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
0        -114.31     34.19                15.0       5612.0          1283.0   
1        -114.47     34.40                19.0       7650.0          1901.0   
2        -114.56     33.69                17.0        720.0           174.0   
3        -114.57     33.64                14.0       1501.0           337.0   
4        -114.57     33.57                20.0       1454.0           326.0   
...          ...       ...                 ...          ...             ...   
16995    -124.26     40.58                52.0       2217.0           394.0   
16996    -124.27     40.69                36.0       2349.0           528.0   
16997    -124.30     41.84                17.0       2677.0           531.0   
16998    -124.30     41.80                19.0       2672.0           552.0   
16999    -124.35     40.54                52.0       1820.0           300.0   

       population  households  median_income  median_house_value  
0          1015.0       472.0         1.4936             66900.0  
1          1129.0       463.0         1.8200             80100.0  
2           333.0       117.0         1.6509             85700.0  
3           515.0       226.0         3.1917             73400.0  
4           624.0       262.0         1.9250             65500.0  
...           ...         ...            ...                 ...  
16995       907.0       369.0         2.3571            111400.0  
16996      1194.0       465.0         2.5179             79000.0  
16997      1244.0       456.0         3.0313            103600.0  
16998      1298.0       478.0         1.9797             85800.0  
16999       806.0       270.0         3.0147             94600.0  

[17000 rows x 9 columns]

In [42]:
housing.columns

Index(['longitude', 'latitude', 'housing_median_age', 'total_rooms',
       'total_bedrooms', 'population', 'households', 'median_income',
       'median_house_value'],
      dtype='object')

In [41]:
titanic = pd.read_csv("/content/train.csv")
titanic

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   NaN        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN        Q  

[891 rows x 12 columns]

In [45]:
titanic.isna().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

In [46]:
titanic["Sex"].value_counts()

male      577
female    314
Name: Sex, dtype: int64

In [47]:
titanic["Pclass"].value_counts()

3    491
1    216
2    184
Name: Pclass, dtype: int64

In [57]:
import numpy as np
arr = np.array([4, 5, 6, 1, 2, 4, 6, 7, 8, 1, 9, 1, 2, 43, 4])
arr[(arr%2 == 0) & (arr>4)]

array([6, 6, 8])

array([6, 6, 8])

In [49]:
#filtering

titanic[titanic["Survived"] == 1]

PassengerId  Survived  Pclass  \
1              2         1       1   
2              3         1       3   
3              4         1       1   
8              9         1       3   
9             10         1       2   
..           ...       ...     ...   
875          876         1       3   
879          880         1       1   
880          881         1       2   
887          888         1       1   
889          890         1       1   

                                                  Name     Sex   Age  SibSp  \
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
8    Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0      0   
9                  Nasser, Mrs. Nicholas (Adele Achem)  female  14.0      1   
..                                                 ...     ...   ...    ...   
875                   Najib, Miss. Adele Kiamie "Jane"  female  15.0      0   
879      Potter, Mrs. Thomas Jr (Lily Alexenia Wilson)  female  56.0      0   
880       Shelley, Mrs. William (Imanita Parrish Hall)  female  25.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
889                              Behr, Mr. Karl Howell    male  26.0      0   

     Parch            Ticket     Fare Cabin Embarked  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
8        2            347742  11.1333   NaN        S  
9        0            237736  30.0708   NaN        C  
..     ...               ...      ...   ...      ...  
875      0              2667   7.2250   NaN        C  
879      1             11767  83.1583   C50        C  
880      1            230433  26.0000   NaN        S  
887      0            112053  30.0000   B42        S  
889      0            111369  30.0000  C148        C  

[342 rows x 12 columns]

In [60]:
titanic[(titanic["Survived"]==1) & (titanic["Sex"]=="male")]

PassengerId  Survived  Pclass                             Name   Sex  \
17            18         1       2     Williams, Mr. Charles Eugene  male   
21            22         1       2            Beesley, Mr. Lawrence  male   
23            24         1       1     Sloper, Mr. William Thompson  male   
36            37         1       3                 Mamee, Mr. Hanna  male   
55            56         1       1                Woolner, Mr. Hugh  male   
..           ...       ...     ...                              ...   ...   
838          839         1       3                  Chip, Mr. Chang  male   
839          840         1       1             Marechal, Mr. Pierre  male   
857          858         1       1           Daly, Mr. Peter Denis   male   
869          870         1       3  Johnson, Master. Harold Theodor  male   
889          890         1       1            Behr, Mr. Karl Howell  male   

      Age  SibSp  Parch  Ticket     Fare Cabin Embarked  
17    NaN      0      0  244373  13.0000   NaN        S  
21   34.0      0      0  248698  13.0000   D56        S  
23   28.0      0      0  113788  35.5000    A6        S  
36    NaN      0      0    2677   7.2292   NaN        C  
55    NaN      0      0   19947  35.5000   C52        S  
..    ...    ...    ...     ...      ...   ...      ...  
838  32.0      0      0    1601  56.4958   NaN        S  
839   NaN      0      0   11774  29.7000   C47        C  
857  51.0      0      0  113055  26.5500   E17        S  
869   4.0      1      1  347742  11.1333   NaN        S  
889  26.0      0      0  111369  30.0000  C148        C  

[109 rows x 12 columns]

In [61]:
housing.mean()

longitude               -119.562108
latitude                  35.625225
housing_median_age        28.589353
total_rooms             2643.664412
total_bedrooms           539.410824
population              1429.573941
households               501.221941
median_income              3.883578
median_house_value    207300.912353
dtype: float64

In [67]:
housing[(housing["total_rooms"]<100) & (housing["housing_median_age"]<10)]

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
113      -115.80     33.26                 2.0         96.0            18.0   
288      -116.76     34.14                 4.0         42.0            10.0   
606      -117.03     33.89                 6.0         78.0            11.0   
1359     -117.17     33.83                 7.0         77.0            12.0   
2273     -117.42     33.89                 4.0         80.0            10.0   
2469     -117.61     34.02                 8.0         63.0             9.0   
2907     -117.76     35.22                 4.0         18.0             3.0   
2990     -117.79     35.21                 4.0          2.0             2.0   
3049     -117.80     33.63                 8.0         32.0             9.0   
9128     -119.01     35.24                 6.0         80.0            16.0   
9260     -119.09     34.22                 8.0         40.0            10.0   
11360    -121.20     37.78                 4.0         58.0            29.0   

       population  households  median_income  median_house_value  
113          30.0        16.0         5.3374             47500.0  
288           9.0         3.0         0.5360             42500.0  
606          27.0        10.0         3.1250            187500.0  
1359         64.0        15.0         4.6000            187500.0  
2273         55.0        13.0         7.7197            193800.0  
2469         25.0         7.0         7.7197            275000.0  
2907          8.0         6.0         1.6250            275000.0  
2990          6.0         2.0         2.3750            137500.0  
3049         26.0        11.0         4.1944            270800.0  
9128         66.0        21.0         3.1250             65000.0  
9260        309.0        16.0         4.0208             52500.0  
11360        79.0        29.0         3.3750            106300.0

In [69]:
housing.describe()

longitude      latitude  housing_median_age   total_rooms  \
count  17000.000000  17000.000000        17000.000000  17000.000000   
mean    -119.562108     35.625225           28.589353   2643.664412   
std        2.005166      2.137340           12.586937   2179.947071   
min     -124.350000     32.540000            1.000000      2.000000   
25%     -121.790000     33.930000           18.000000   1462.000000   
50%     -118.490000     34.250000           29.000000   2127.000000   
75%     -118.000000     37.720000           37.000000   3151.250000   
max     -114.310000     41.950000           52.000000  37937.000000   

       total_bedrooms    population    households  median_income  \
count    17000.000000  17000.000000  17000.000000   17000.000000   
mean       539.410824   1429.573941    501.221941       3.883578   
std        421.499452   1147.852959    384.520841       1.908157   
min          1.000000      3.000000      1.000000       0.499900   
25%        297.000000    790.000000    282.000000       2.566375   
50%        434.000000   1167.000000    409.000000       3.544600   
75%        648.250000   1721.000000    605.250000       4.767000   
max       6445.000000  35682.000000   6082.000000      15.000100   

       median_house_value  
count        17000.000000  
mean        207300.912353  
std         115983.764387  
min          14999.000000  
25%         119400.000000  
50%         180400.000000  
75%         265000.000000  
max         500001.000000

'Braund'

In [82]:
lname= []
for i in range(0, len(titanic["Name"])):
    lname.append(titanic["Name"][i].split(",")[0])

titanic["last_name"] = lname
titanic

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  last_name  
0        0         A/5 21171   7.2500   NaN        S     Braund  
1        0          PC 17599  71.2833   C85        C    Cumings  
2        0  STON/O2. 3101282   7.9250   NaN        S  Heikkinen  
3        0            113803  53.1000  C123        S   Futrelle  
4        0            373450   8.0500   NaN        S      Allen  
..     ...               ...      ...   ...      ...        ...  
886      0            211536  13.0000   NaN        S   Montvila  
887      0            112053  30.0000   B42        S     Graham  
888      2        W./C. 6607  23.4500   NaN        S   Johnston  
889      0            111369  30.0000  C148        C       Behr  
890      0            370376   7.7500   NaN        Q     Dooley  

[891 rows x 13 columns]

In [84]:
titanic["last_name"] = [titanic["Name"][i].split(",")[0] for i in range(0, len(titanic["Name"]))]
titanic

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  last_name  
0        0         A/5 21171   7.2500   NaN        S     Braund  
1        0          PC 17599  71.2833   C85        C    Cumings  
2        0  STON/O2. 3101282   7.9250   NaN        S  Heikkinen  
3        0            113803  53.1000  C123        S   Futrelle  
4        0            373450   8.0500   NaN        S      Allen  
..     ...               ...      ...   ...      ...        ...  
886      0            211536  13.0000   NaN        S   Montvila  
887      0            112053  30.0000   B42        S     Graham  
888      2        W./C. 6607  23.4500   NaN        S   Johnston  
889      0            111369  30.0000  C148        C       Behr  
890      0            370376   7.7500   NaN        Q     Dooley  

[891 rows x 13 columns]

In [90]:
#recommended

def get_last_name(name):
    res = name.split(",")[0].lower()

    return res

titanic["Name"].apply(get_last_name)

0         braund
1        cumings
2      heikkinen
3       futrelle
4          allen
         ...    
886     montvila
887       graham
888     johnston
889         behr
890       dooley
Name: Name, Length: 891, dtype: object

In [89]:
titanic["Name"].apply(lambda name: name.split(",")[0])

0         Braund
1        Cumings
2      Heikkinen
3       Futrelle
4          Allen
         ...    
886     Montvila
887       Graham
888     Johnston
889         Behr
890       Dooley
Name: Name, Length: 891, dtype: object

In [102]:

titanic["gender_numeric"] = (titanic["Sex"] == "female").astype(int)
titanic

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  last_name  \
0        0         A/5 21171   7.2500   NaN        S     Braund   
1        0          PC 17599  71.2833   C85        C    Cumings   
2        0  STON/O2. 3101282   7.9250   NaN        S  Heikkinen   
3        0            113803  53.1000  C123        S   Futrelle   
4        0            373450   8.0500   NaN        S      Allen   
..     ...               ...      ...   ...      ...        ...   
886      0            211536  13.0000   NaN        S   Montvila   
887      0            112053  30.0000   B42        S     Graham   
888      2        W./C. 6607  23.4500   NaN        S   Johnston   
889      0            111369  30.0000  C148        C       Behr   
890      0            370376   7.7500   NaN        Q     Dooley   

     gender_numeric  
0                 0  
1                 1  
2                 1  
3                 1  
4                 0  
..              ...  
886               0  
887               1  
888               1  
889               0  
890               0  

[891 rows x 14 columns]

In [108]:
titanic["Embarked"].value_counts()

S    644
C    168
Q     77
Name: Embarked, dtype: int64

In [113]:
titanic["Embarked"].map({"S":0, "C":1, "Q":2})

0      0.0
1      1.0
2      0.0
3      0.0
4      0.0
      ... 
886    0.0
887    0.0
888    0.0
889    1.0
890    2.0
Name: Embarked, Length: 891, dtype: float64

In [115]:
titanic["Sex"].map({"male":0, "female":1})

0      0
1      1
2      1
3      1
4      0
      ..
886    0
887    1
888    1
889    0
890    0
Name: Sex, Length: 891, dtype: int64

In [118]:
titanic[titanic["Embarked"].isnull()==True]

PassengerId  Survived  Pclass                                       Name  \
61            62         1       1                        Icard, Miss. Amelie   
829          830         1       1  Stone, Mrs. George Nelson (Martha Evelyn)   

        Sex   Age  SibSp  Parch  Ticket  Fare Cabin Embarked last_name  \
61   female  38.0      0      0  113572  80.0   B28      NaN     Icard   
829  female  62.0      0      0  113572  80.0   B28      NaN     Stone   

     gender_numeric  
61                1  
829               1

In [128]:
print(titanic[titanic["Sex"]=="male"]["Age"].mean())
print(titanic[titanic["Sex"]=="female"]["Age"].mean())


30.72664459161148
27.915708812260537


In [132]:
titanic.groupby(["Sex"]).mean()[["Age", "Fare"]]

Age       Fare
Sex                         
female  27.915709  44.479818
male    30.726645  25.523893

In [135]:
titanic.groupby(["Embarked"]).mean()[["Age", "Fare"]]


Age       Fare
Embarked                      
C         30.814769  59.954144
Q         28.089286  13.276030
S         29.445397  27.079812

In [138]:
titanic[(titanic["Sex"]=="male") & titanic["Survived"]==1]["Age"].mean()
titanic[(titanic["Sex"]=="male") & titanic["Survived"]==0]["Age"].mean()
titanic[(titanic["Sex"]=="female") & titanic["Survived"]==1]["Age"].mean()
titanic[(titanic["Sex"]=="female") & titanic["Survived"]==0]["Age"].mean()


27.276021505376345

In [142]:
titanic.groupby(["Sex", "Survived"]).mean()

PassengerId    Pclass        Age     SibSp     Parch  \
Sex    Survived                                                         
female 0          434.851852  2.851852  25.046875  1.209877  1.037037   
       1          429.699571  1.918455  28.847716  0.515021  0.515021   
male   0          449.121795  2.476496  31.618056  0.440171  0.207265   
       1          475.724771  2.018349  27.276022  0.385321  0.357798   

                      Fare  gender_numeric  
Sex    Survived                             
female 0         23.024385             1.0  
       1         51.938573             1.0  
male   0         21.960993             0.0  
       1         40.821484             0.0

In [144]:
titanic.groupby(["Sex", "Survived", "Pclass"]).mean()


PassengerId        Age     SibSp     Parch  \
Sex    Survived Pclass                                               
female 0        1        325.000000  25.666667  0.666667  1.333333   
                2        423.500000  36.000000  0.500000  0.166667   
                3        440.375000  23.818182  1.291667  1.097222   
       1        1        473.967033  34.939024  0.549451  0.428571   
                2        444.785714  28.080882  0.485714  0.642857   
                3        359.083333  19.329787  0.500000  0.500000   
male   0        1        413.623377  44.581967  0.272727  0.259740   
                2        454.010989  33.369048  0.307692  0.142857   
                3        456.750000  27.255814  0.523333  0.213333   
       1        1        527.777778  36.248000  0.377778  0.311111   
                2        415.588235  16.022000  0.529412  0.647059   
                3        447.638298  22.274211  0.340426  0.297872   

                              Fare  gender_numeric  
Sex    Survived Pclass                              
female 0        1       110.604167             1.0  
                2        18.250000             1.0  
                3        19.773093             1.0  
       1        1       105.978159             1.0  
                2        22.288989             1.0  
                3        12.464526             1.0  
male   0        1        62.894910             0.0  
                2        19.488965             0.0  
                3        12.204469             0.0  
       1        1        74.637320             0.0  
                2        21.095100             0.0  
                3        15.579696             0.0

In [123]:
titanic.groupby(["Survived"]).mean()

PassengerId    Pclass        Age     SibSp     Parch       Fare  \
Survived                                                                    
0          447.016393  2.531876  30.626179  0.553734  0.329690  22.117887   
1          444.368421  1.950292  28.343690  0.473684  0.464912  48.395408   

          gender_numeric  
Survived                  
0               0.147541  
1               0.681287

In [27]:
# We load fake Company data in a DataFrame
data = pd.read_csv('/content/drive/MyDrive/devtown-10922/DevTown/Python-for-Data-Science--master/fake_company.csv')

data

Year     Name Department  Age  Salary
0  1990    Alice         HR   25   50000
1  1990      Bob         RD   30   48000
2  1990  Charlie      Admin   45   55000
3  1991    Alice         HR   26   52000
4  1991      Bob         RD   31   50000
5  1991  Charlie      Admin   46   60000
6  1992    Alice      Admin   27   60000
7  1992      Bob         RD   32   52000
8  1992  Charlie      Admin   28   62000

We see that the data contains information for the year 1990 through 1992. For each year we see name of the employees, the department they work for, their age, and their annual salary. Now, let's use the ```.groupby()``` method to get information.

Let's calculate how much money the company spent in salaries each year. To do this, we will group the data by Year using the ```.groupby()``` method and then we will add up the salaries of all the employees by using the ```.sum()``` method.

In [ ]:
# We display the total amount of money spent in salaries each year
data.groupby(['Year'])['Salary'].sum()

Year
1990    153000
1991    162000
1992    174000
Name: Salary, dtype: int64

We see that the company spent a total of 153,000 dollars in 1990, 162,000 in 1991, and 174,000 in 1992.

Now, let's suppose I want to know what was the average salary for each year. In this case, we will group the data by Year using the ```.groupby()``` method, just as we did before, and then we use the ```.mean()``` method to get the average salary. Let's see how this works

In [ ]:
# We display the average salary per year
data.groupby(['Year'])['Salary'].mean()

Year
1990    51000
1991    54000
1992    58000
Name: Salary, dtype: int64

We see that the average salary in 1990 was 51,000 dollars, 54,000 in 1991, and 58,000 in 1992.

Now let's see how much did each employee get paid in those three years. In this case, we will group the data by Name using the ```.groupby()``` method and then we will add up the salaries for each year. Let's see the result

In [ ]:
# We display the salary distribution per department per year.
data.groupby(['Year', 'Department'])['Salary'].sum()

Year  Department
1990  Admin          55000
      HR             50000
      RD             48000
1991  Admin          60000
      HR             52000
      RD             50000
1992  Admin         122000
      RD             52000
Name: Salary, dtype: int64

We see that in 1990 the Admin department paid 55,000 dollars in salaries,the HR department paid 50,000, and the RD department 48,0000. While in 1992 the Admin department paid 122,000 dollars in salaries and the RD department paid 52,000.